scripts/upgrade/database.php

```diff
@@ -226,40 +227,49 @@ function updateIntoTable( $updateTable, $fromTable, $setClause, $whereClause )
-    $tbl_user      = call_user_func('User::getDBTable');
-
-    $if_not_exists = '';
-    if( $dbtype == 'mysql' ) {
-        $if_not_exists = ' IF NOT EXISTS ';
-    }
-
+	$tbl_auth       = call_user_func('Authentication::getDBTable');
+	$tbl_user       = array(
+		'tbName' => call_user_func('User::getDBTable'),
+		'indexName' => 'UserPreferences_authid',
+		'reference' => $tbl_auth
+	);
+	$tbl_transfers    = array(
+		'tbName' => call_user_func('Transfer::getDBTable'),
+		'indexName' => 'Transfers_userid',
+		'reference' => $tbl_user
+	);
+	$tbl_guests      = array(
+		'tbName' => call_user_func('Guest::getDBTable'),
+		'indexName' => 'Guests_userid',
+		'reference' => $tbl_user
+	);
+	$tbl_clientlog    = array(
+		'tbName' => call_user_func('ClientLog::getDBTable'),
+		'indexName' => 'ClientLog_userid',
+		'reference' => $tbl_user
+	);
+
+	$tbls = array($tbl_user, $tbl_transfers, $tbl_guests, $tbl_clientlog);
     $a = array();
-    array_push( $a,'ALTER TABLE '.$tbl_user.     ' ADD CONSTRAINT UserPreferences_authid FOREIGN KEY ' . $if_not_exists . ' (authid) REFERENCES '.$tbl_auth.' (id)  on delete cascade on update restrict ;');
-    array_push( $a, 'ALTER TABLE '.$tbl_transfers.' ADD CONSTRAINT Transfers_userid       FOREIGN KEY ' . $if_not_exists . ' (userid) REFERENCES '.$tbl_user.' (id)  on delete cascade on update restrict ;');
-    array_push( $a, 'ALTER TABLE '.$tbl_guests.   ' ADD CONSTRAINT Guests_userid          FOREIGN KEY ' . $if_not_exists . ' (userid) REFERENCES '.$tbl_user.' (id)  on delete cascade on update restrict ;');
-    array_push( $a, 'ALTER TABLE '.$tbl_clientlog.' ADD CONSTRAINT ClientLog_userid       FOREIGN KEY ' . $if_not_exists . ' (userid) REFERENCES '.$tbl_user.' (id)  on delete cascade on update restrict ;');
-
-    // There is no if not exists for psql so we just make
-    // all the fks and ignore the duplicate error coming back.
-    // Other errors are left to hit top level as fatal
-    foreach ($a as $key => $sql)
-    {
-        try {
-            echo $sql . "\n";
-            DBI::exec( $sql );
-        } catch(DBIDuplicateException $e ) {
-            echo "... already there!\n";
-        }
-    }
+
+	foreach ( $tbls as $tbl ) {
+		// We check the presence of the foreign key
+		if( $dbtype == 'pgsql' ) {
+			$checkQuery = DBI::query("SELECT COUNT(1) FROM INFORMATION_SCHEMA.table_constraints WHERE constraint_name='".$tbl['indexName']."' AND table_name='".$tbl['tbName']."';");
+		}
+		else {
+			$checkQuery = DBI::query("SELECT COUNT(1) indexExists FROM INFORMATION_SCHEMA.STATISTICS WHERE table_schema='".Config::get('db_database')."' AND table_name='".$tbl['tbName']."' AND index_name='".$tbl['indexName']."';");
+		}
+
+		if ( $checkQuery->rowCount() === 0) {
+			array_push( $a,'ALTER TABLE '.$tbl['tbName'].' ADD CONSTRAINT '.$tbl['indexName'].' FOREIGN KEY (authid) REFERENCES '.$tbl['reference'].' (id)  on delete cascade on update restrict ;');
```

Comment:
Note that the fk is authid on the first case and userid on the next 3. This SQL statements doesn't run on postgresql fir Transfers because there is no authid column in that table. 

It seems this got through the CI because the foreach iterating over $a and actually executing the SQL to create the FK is removed from the PR.

Reply:
'.$tbl['reference'].' in the SQL will give an Array instead of the table name.